src/main/java/com/sapient/model/service/BudgetCategoryService.java

```diff
@@ -0,0 +1,124 @@
+package com.sapient.model.service;
+
+import com.sapient.exception.*;
+import com.sapient.model.beans.Budget;
+import com.sapient.model.beans.BudgetCategory;
+import com.sapient.model.beans.User;
+import com.sapient.model.dao.BudgetCategoryRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+public class BudgetCategoryService {
+
+    @Autowired
+    private BudgetCategoryRepository budgetCategoryDao;
+
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    private CategoryService categoryService;
+
+    @Autowired
+    private BudgetService budgetService;
+
+    public BudgetCategory createBudgetCategory(String passwordHash, Float amount, Integer categoryId, Integer budgetId)
+            throws NotAuthorizedException, BudgetCategoryTakenException, CategoryNotFoundException,
+            RecordNotFoundException {
+        User user;
+        try{
+            user = userService.getUserByPasswordHash(passwordHash);
+        }catch (UserNotFoundException e){
+            throw new NotAuthorizedException("You are not authorized to create a budgetCategory");
+        }
+
+        if(budgetCategoryTaken(categoryId, budgetId, user.getId())){
+            throw new BudgetCategoryTakenException("BudgetCategory for budget: "+ budgetId + ", category: " +
+                    categoryId + " is already created for user " + user.getUsername() );
+        }
+
+        BudgetCategory budgetCategory = new BudgetCategory();
+
+        Budget budget = budgetService.getBudget(passwordHash, budgetId);
+
+        budgetCategory.setCategory(categoryService.getCategory(passwordHash, categoryId));
+        budgetCategory.setBudget(budget);
+        budgetCategory.setAmount(amount);
+        budgetCategory.setUser(user);
+
+        budgetCategoryDao.save(budgetCategory);
+        return budgetCategory;
+    }
+
+    public BudgetCategory updateBudgetCategory(String passwordHash, Integer id, Float amount) throws NotAuthorizedException {
+        User user;
+        BudgetCategory budgetCategory;
+        try{
+            user = userService.getUserByPasswordHash(passwordHash);
+            budgetCategory = getBudgetCategory(passwordHash, id);
+        }catch (UserNotFoundException | RecordNotFoundException e){
+            throw new NotAuthorizedException("You are not authorized to update this budgetCategory");

```

Comment:
If user is a valid user but budget category does not exist, throw a RecordNotFoundException instead of a NotAuthorizedException. If budget category does exist but belongs to another user, also throw a RecordNotFoundException.